proxy/__init__.py

```diff
@@ -55,15 +55,38 @@ def __init__(self, config: ProxyCacheConfig, repository: str, validation: bool =
             config.upstream_registry_hostname,
             config.upstream_registry_hostname,
         )
-        url = f"https://{hostname}"
-        if config.insecure:
-            url = f"http://{hostname}"
 
-        self.base_url = url
         self._session = requests.Session()
+
+        self.base_url = self._get_valid_upstream_registry(hostname)
         self._repo = repository
         self._authorize(self._credentials(), force_renewal=self._validation)
-        # flag used for validating Proxy cache config before saving to db
+        # flag used for validating Proxy cache config before saving to db    
+
+    def _get_valid_upstream_registry(self, hostname: str) -> str :
+        url = f"https://{hostname}"
+        try:
+            resp = self._session.get(f"{url}/v2")
```

Comment:
use `self._safe_request` everywhere in this function. there's an example of it being used in `_authorize`